### ISSUE TITLE
AverageImage or mean of the training images ?

### ISSUE THREAD

**Issue by kbenahmed89** [vlfeat/matconvnet#485]
HI,

I'm tuning the pre-trained network imagenet-vgg-f,while preprocessing the training images,  should I subtract the averageImage provided by the network  or subtract the mean of all images or go with every image and subtract its mean from the original one?

what way is the correct way? and in what cases I need to use each of these ways?

how about when I want to test the tuned network, what way should I use?

thanks.

**Comment by albanie** [comment]
The average image contained in `meta.normalization.averageImage` attached to `imagenet-vgg-f.mat` is the average across all images that the network was originally trained on.  

For both fine tuning and testing, you should subtract this average from any new images you feed into the network.

**Comment by kbenahmed89** [comment]
Hi

thanks for your reply, but for fine tuning I'm using images that have nothing to do with the image that the network was trained on. does it make sense to subtract the mean of imagenet images even though I'm tuning with MRI medical images?

**Comment by albanie** [comment]
In practice, the features trained from imagenet tend to generalize pretty effectively to new sets of images so subtraction by the original mean works well as a starting point for fine tuning.  

On the other hand, since you are working with MRI images it may be that your target dataset has dramatically different characteristics to the original training data - you may find that subtracting the mean of the new data gives a better initialization. 

In [this paper](https://hal.inria.fr/hal-01162880/file/recognizing-cardiac-magnetic-submission.pdf) the authors similarly fine tuned a network that had been pretrained on imagenet and achieve good classification results on an MRI dataset after subtracting the mean of their *target MRI dataset*, so I'd say it's worth trying both to see what works best on your data.

**Comment by kbenahmed89** [comment]
thanks, but is there any theoretical explanation of why should we subtract this average ? is there a paper where you explain why and how it affects the network?

**Comment by albanie** [comment]
The intuition behind zero centering the data (i.e. subtracting the image mean) before inputs are fed into the network is explained in Section 4.3 of [Efficient Backprop](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf).  When you run the imagenet example, mean image subtraction is performed as a preprocessing step on the training and validation data (this is done at the end of [cnn_imagenet_get_batch](https://github.com/vlfeat/matconvnet/blob/master/examples/imagenet/cnn_imagenet_get_batch.m)).

When it comes to fine tuning, if you want to start from *exactly* the same features that were used during the training on the imagenet example, you need to perform the same preprocessing. To achieve this, you must subtract the same mean image that was calculated from the original training data.  This won't zero-center your new data, it simply means that the same features will be produced by the network.

Conversely, If you instead subtract the mean of your target data, your data will become zero centered, but you will no longer be using exactly the same features used during the imagenet pretraining. The relative importance of using *zero centered data* vs *using the original imagenet features* will depend on how much fine tuning occurs and how much your target dataset differs from the imagenet data.  Unfortunately I'm not aware of any theoretical explanations that cover this point  but if you find something, I'd be very happy to hear about it :)

**Comment by kbenahmed89** [comment]
thank you so much for the explanation, I just wanted to make sure that I'm doing this subtraction right please, if I wanted to subtract the mean of all my data I use this code: 

imageMean = mean(imdb.images.data(:)) ;
imdb.images.data = imdb.images.data - imageMean ;

and if I wanted to subtract the mean of each image I use this code:

for i=1:44
    im= imdb.images.data(:,:,:,i);
    themean = mean(im(:));
    imdb.images.data(:,:,:,i)= imdb.images.data(:,:,:,i) - themean;
end

is this the right way ? I got the mean as just one value and not a 224x224 image like the net.meta.normalization.averageimage? is this correct ?

**Comment by albanie** [comment]
The goal is to subtract the mean of all the *training* images, rather than the mean of each individual image.  The following code makes some assumptions about your `imdb` structure, so can be modified if needed but it should give you the basic idea:

* It assumes `imdb.images.data` is a four dimensional array of size `H x W x D x N` where `N` is the total number of images.
* It assumes that `imdb.images.set` is an `N x 1` array of integers in which `1` represents that an image is a member of the training set and `2` represents a member of the validation set.

        % find the indices of the training images in the dataset
        trainIdx = find(imdb.images.set == 1) ;

        % compute the mean across all training images 
        imageMean = mean(imdb.images.data(:,:,:,trainIdx), 4);

        % loop over ALL images to subtract the training image mean
        for i = 1:size(imdb.images.data, 4)
            imdb.images.data(:,:,:,i) = imdb.images.data(:,:,:,i) - imageMean;
        end

In the example given above, the `imageMean` is an H x W x D array.  This code assumes that you intend to subtract the mean of your *target data* (as opposed to the original imagenet data as we discussed, which can be done by replacing `imageMean` in the for loop by  `meta.normalization.averageImage` and modifying the dimensions to make the subtraction work).
